### PR TITLE
Integration of boundary conditions of first-type

### PR DESCRIPTION
**Pull request by TomFischer**
- Integration of boundary conditions of first-type into the system of linear equations using the `CRSMatrix` as system matrix
- Converter to use this integration together with the LIS data structures

### CODE REVIEW

File: MathLib/LinAlg/Lis/LisTools.cpp [hunk 1]
```diff
@@ -14,34 +14,86 @@
...
-    const std::size_t n_bc = vec_knownX_id.size();
-    for (std::size_t i_bc=0; i_bc<n_bc; i_bc++) {
-        const std::size_t rowId = vec_knownX_id[i_bc];
-        const double x = vec_knownX_x[i_bc];
-        //A(k, k) = penalty
-        A.setValue(rowId, rowId, penalty);
-        //b(k) = x*penalty
-        b.set(rowId, x*penalty);
-    }
+MathLib::CRSMatrix<double, typename LisMatrix::IndexType>* lis2crs(LisMatrix &a)
```
Comment (endJunction) [inline MathLib/LinAlg/Lis/LisTools.cpp]: Some comments for the main steps of the algorithm would be helpful. (Yes, it's just a deep copy but maybe there is some special treatment like sorting or indices counting from 1...)

Comment (TomFischer) [inline MathLib/LinAlg/Lis/LisTools.cpp]: Added documentation.

File: MathLib/LinAlg/Lis/LisTools.cpp [hunk 2]
```diff
@@ -14,34 +14,89 @@
...
-        //A(k, k) = penalty
-        A.setValue(rowId, rowId, penalty);
-        //b(k) = x*penalty
-        b.set(rowId, x*penalty);
-    }
+	using IndexType = LisMatrix::IndexType;
+
+	LIS_MATRIX &A = a.getRawMatrix();
+
+	IndexType const n(A->n);
```
Comment (chleh) [inline MathLib/LinAlg/Lis/LisTools.cpp]: Could you add a hint that this is the number of rows, please.

File: MathLib/LinAlg/Sparse/CRSTools-impl.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	typename VEC_T::IndexType const*const jA(mat->getColIdxArray()); // col idx
+	double * entries(const_cast<double*>(mat->getEntryArray()));
+
+	// set row entries, except the diagonal entry, to zero
+	for (std::size_t r(0); r<rows.size(); ++r) {
+		auto const row = rows[r];
+		for (unsigned j(iA[row]); j<iA[row+1]; ++j) {
+			if (jA[j] == row) {
+				// rhs_{row} = a_{row,row} * val
+				rhs.set(row, entries[j]*vals[r]);
```
Comment (chleh) [inline MathLib/LinAlg/Sparse/CRSTools-impl.h]: What if a matrix has a zero on the diagonal?
In this case I would set the diagonal to one for the moment and leave a better solution for later.

Comment (wenqing) [inline MathLib/LinAlg/Sparse/CRSTools-impl.h]: In addition, for this serial solver, it it better to also set off-diagonal entries of the corresponding column be zero. Under such consideration,  symmetrical matrix is still symmetry after calling this function.

Comment (TomFischer) [inline MathLib/LinAlg/Sparse/CRSTools-impl.h]: @wenqing What do you mean by "this serial solver".  The function at hand does integrate boundary conditions of first-type. The algorithm works as follows:
1. Compute the transposed matrix and set the non-zero column entries (except the diagonal entry) for the given rows to zero and integrate the information into the rhs.
2. Compute the transposed matrix of the transposed matrix (= original matrix) and again, remove the non-zero column entries.
The symmetry of the matrix is preserved. Did I made a mistake?

Comment (wenqing) [inline MathLib/LinAlg/Sparse/CRSTools-impl.h]: It's clear to me now. Thanks for explanation.

File: MathLib/LinAlg/Lis/LisTools.cpp (hunk 2 above)
Comment (TomFischer) [inline MathLib/LinAlg/Lis/LisTools.cpp]: Renamed variable and added docu.

File: MathLib/LinAlg/Sparse/CRSTools-impl.h (hunk 3 above)
Comment (TomFischer) [inline MathLib/LinAlg/Sparse/CRSTools-impl.h]: @chleh I am unsure what it means when the diagonal entries equals zero. I wonder which conditions have to be fulfilled if this happens? Maybe this is the consequence of an inconsistent first-type boundary condition?! If there are no objections I would discuss this separately from this PR.

Comment (chleh) [inline MathLib/LinAlg/Sparse/CRSTools-impl.h]: In general matrices can have zeroes on the diagonal. Maybe that's rather unlikely for FEM. But can you guarantee that there are never zeroes even for indefinite matrices, or when there are multiple primary variables or when you are solving the equations arising from a transient or nonlinear problem?
Does anybody else see those issues?

**Comment by chleh** [comment]
:+1:
Please make an issue from my last comment (unless it is obviously irrelevant)

**Comment by endJunction** [comment]
Simply check for A(row,row) == 0 and set to 1 in that case.

**Pull request merged**